### DISCUSSION TITLE
Issues with setting up and multiple languages

### DISCUSSION THREAD

**Discussion by RoseSamaras** [naptha/tesseract.js#968] [Q&A]
Hello people, I am Rose from Greece, I am building a business card reader, and I am having lots of issues with language packs, I am a noob and I use a lot of claude to help me, but i cant make it work, results are really bad, and it doesnt read Greek at all, can anyone help?

BusinessCardScanner.jsx
import React, { useState, useRef, useEffect } from 'react';
import { 
  Button, 
  VStack, 
  Image, 
  useToast, 
  Box, 
  Text,
  HStack,
  Center,
  Spinner,
  Progress,
  IconButton,
  Icon
} from '@chakra-ui/react';
import Webcam from 'react-webcam';
import jsQR from 'jsqr';
import { parseVCard, parseBusinessCard } from '../utils/dataExtraction';
import OCRService from '../services/ocrService';
import { MdCameraAlt, MdRefresh } from 'react-icons/md';
import { RiQrScanLine, RiCameraLine, RiRefreshLine } from 'react-icons/ri';

const CAPTURE_OPTIONS = {
  width: 1920,
  height: 1080,
  facingMode: 'environment'
};

// BusinessCardScanner.jsx
const BusinessCardScanner = ({ onExtractedData }) => {
  const [mode, setMode] = useState('qr'); // 'qr' or 'card'
  const [capturedImage, setCapturedImage] = useState(null);
  const [isProcessing, setIsProcessing] = useState(false);
  const [extractedText, setExtractedText] = useState([]);
  const [scanProgress, setScanProgress] = useState(0);
  const [isInitialized, setIsInitialized] = useState(false);

  
  const webcamRef = useRef(null);
  const qrIntervalRef = useRef(null);
  const canvasRef = useRef(document.createElement('canvas'));
  const toast = useToast();

  // QR Code scanning
  useEffect(() => {
    if (mode !== 'qr' || !webcamRef.current || isProcessing) return;

    qrIntervalRef.current = setInterval(() => {
      const video = webcamRef.current?.video;
      if (!video || video.readyState !== video.HAVE_ENOUGH_DATA) return;

      const canvas = document.createElement('canvas');
      const context = canvas.getContext('2d', { willReadFrequently: true });
      
      canvas.width = video.videoWidth;
      canvas.height = video.videoHeight;
      context.drawImage(video, 0, 0);

      try {
        const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
        const code = jsQR(imageData.data, imageData.width, imageData.height, {
          inversionAttempts: "attemptBoth"
        });

        if (code) {
          handleQRCode(code.data);
        }
      } catch (error) {
        console.error('QR scanning error:', error);
      }
    }, 100);

    return () => {
      if (qrIntervalRef.current) {
        clearInterval(qrIntervalRef.current);
      }
    };
  }, [mode, isProcessing]);

  // Initialize OCR
  useEffect(() => {
    let mounted = true;

    const initialize = async () => {
      try {
        await OCRService.getWorker();
        if (mounted) {
          setIsInitialized(true);
          toast({
            title: "Ready",
            description: "Scanner initialized successfully",
            status: "success",
            duration: 2000,
          });
        }
      } catch (error) {
        console.error('Scanner initialization error:', error);
        if (mounted) {
          toast({
            title: "Initialization Error",
            description: "Failed to initialize scanner. Please refresh the page.",
            status: "error",
            duration: 5000,
            isClosable: true,
          });
        }
      }
    };

    initialize();

    return () => {
      mounted = false;
      OCRService.terminate();
    };
  }, [toast]);

  const handleQRCode = async (data) => {
    try {
      setIsProcessing(true);
      
      if (data.startsWith('BEGIN:VCARD')) {
        const { vcardData, mappings } = parseVCard(data);
        setExtractedText(vcardData);
        onExtractedData({
          type: 'vcard',
          vcardData,
          mappings,
          raw: data
        });

        toast({
          title: "Success",
          description: "Contact information extracted",
          status: "success",
          duration: 2000,
        });
      } else {
        toast({
          title: "Invalid Format",
          description: "QR code must contain contact information",
          status: "warning",
          duration: 3000,
        });
      }
    } catch (error) {
      console.error('QR processing error:', error);
      toast({
        title: "Error",
        description: "Failed to process QR code",
        status: "error",
        duration: 3000,
      });
    } finally {
      setIsProcessing(false);
    }
  };

  const handleCapture = async () => {
    if (!webcamRef.current || !isInitialized) {
      toast({
        title: "Error",
        description: "Camera or scanner not initialized",
        status: "error",
        duration: 3000,
      });
      return;
    }

    try {
      setIsProcessing(true);
      setScanProgress(20);
      
      const imageSrc = webcamRef.current.getScreenshot();
      if (!imageSrc) {
        throw new Error('Failed to capture image');
      }
      
      // Process and enhance the image
      const processedImage = await preprocessImage(imageSrc);
      setCapturedImage(processedImage);
      setScanProgress(40);
      
      // Perform OCR
      const text = await OCRService.recognize(processedImage);
      setScanProgress(80);
      
      if (!text || text.trim().length === 0) {
        throw new Error('No text was extracted from the image');
      }

      // Process the extracted text
      const lines = text
        .split('\n')
        .map(line => line.trim())
        .filter(line => line.length > 0)
        .map(line => line.replace(/\s+/g, ' ')); // Clean up extra spaces

      setExtractedText(lines);
      onExtractedData({
        type: 'ocr',
        data: lines,
        raw: text
      });

      toast({
        title: "Success",
        description: "Business card scanned successfully",
        status: "success",
        duration: 2000,
      });
    } catch (error) {
      console.error('Card processing error:', error);
      toast({
        title: "Error",
        description: error.message || 'Failed to process business card',
        status: "error",
        duration: 3000,
      });
    } finally {
      setIsProcessing(false);
      setScanProgress(0);
    }
  };
  
  const preprocessImage = (imageSrc) => {
    return new Promise((resolve) => {
      const img = document.createElement('img');
      img.onload = () => {
        const canvas = canvasRef.current;
        const ctx = canvas.getContext('2d');
        
        // Set to a higher resolution
        canvas.width = img.width * 2;
        canvas.height = img.height * 2;
        
        // Clear previous content
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        
        // Apply image processing
        ctx.filter = 'contrast(1.2) brightness(1.1)';
        ctx.drawImage(img, 0, 0, canvas.width, canvas.height);
        
        // Get processed image
        resolve(canvas.toDataURL('image/jpeg', 0.95));
      };
      img.src = imageSrc;
    });
  };

  const handleRetake = () => {
    setCapturedImage(null);
    setExtractedText([]);
    setMode('qr');
    setIsProcessing(false);
    setScanProgress(0);
  };

  const toggleMode = () => {
    setMode(prev => prev === 'qr' ? 'card' : 'qr');
    setCapturedImage(null);
    setExtractedText([]);
  };

  return (
    <VStack spacing={4} w="full" maxW="800px" mx="auto">
      <HStack spacing={4} w="full" justify="center">
        <Button
          colorScheme={mode === 'qr' ? 'green' : 'gray'}
          onClick={() => setMode('qr')}
          leftIcon={<Icon as={RiQrScanLine} />}
        >
          QR Code
        </Button>
        <Button
          colorScheme={mode === 'card' ? 'blue' : 'gray'}
          onClick={() => setMode('card')}
          leftIcon={<Icon as={RiCameraLine} />}
        >
          Paper Card
        </Button>
      </HStack>

      <Box position="relative" width="100%" aspectRatio={4/3}>
        {!capturedImage ? (
          <>
            <Webcam
              audio={false}
              ref={webcamRef}
              screenshotFormat="image/jpeg"
              videoConstraints={CAPTURE_OPTIONS}
              style={{
                width: '100%',
                height: '100%',
                objectFit: 'cover'
              }}
            />
            {mode === 'qr' && (
              <Box
                position="absolute"
                top="50%"
                left="50%"
                transform="translate(-50%, -50%)"
                border="3px solid #22c35e"
                width="280px"
                height="280px"
                borderRadius="20px"
                boxShadow="0 0 0 9999px rgba(0, 0, 0, 0.5)"
              />
            )}
          </>
        ) : (
          <Image 
            src={capturedImage} 
            alt="Captured image" 
            objectFit="contain"
            w="100%"
            h="100%"
          />
        )}

        {isProcessing && (
          <Center
            position="absolute"
            top="0"
            left="0"
            right="0"
            bottom="0"
            bg="blackAlpha.300"
          >
            <Spinner size="xl" color="blue.500" thickness="4px" />
          </Center>
        )}
      </Box>

      {scanProgress > 0 && scanProgress < 100 && (
        <Progress 
          value={scanProgress} 
          width="100%" 
          size="sm" 
          colorScheme="blue" 
          isAnimated
        />
      )}

      <HStack spacing={4} justify="center" w="full">
        {!capturedImage ? (
          mode === 'card' && (
            <Button
              colorScheme="blue"
              onClick={handleCapture}
              isDisabled={isProcessing}
              size="lg"
              leftIcon={<Icon as={RiCameraLine} />}
            >
              Capture Card
            </Button>
          )
        ) : (
          <Button
            colorScheme="purple"
            onClick={handleRetake}
            isDisabled={isProcessing}
            size="lg"
            leftIcon={<Icon as={RiRefreshLine} />}
          >
            Retake
          </Button>
        )}
      </HStack>

      {extractedText.length > 0 && (
        <Box mt={4} p={4} borderRadius="md" borderWidth={1} w="full">
          <Text fontWeight="bold" mb={2}>Extracted Data:</Text>
          {extractedText.map((line, index) => (
            <Text key={index}>{line}</Text>
          ))}
        </Box>
      )}
    </VStack>
  );
};

export default BusinessCardScanner;

OCR Services
import { createWorker } from 'tesseract.js';

class OCRService {
  static worker = null;
  static isInitializing = false;

  static async getWorker() {
    if (OCRService.worker) {
      return OCRService.worker;
    }

    if (OCRService.isInitializing) {
      while (OCRService.isInitializing) {
        await new Promise(resolve => setTimeout(resolve, 100));
      }
      return OCRService.worker;
    }

    try {
      OCRService.isInitializing = true;

      const worker = await createWorker({
        langPath: '/tessdata',
        logger: m => console.log(m),
      });

      // Initialize with both English and Greek
      await worker.loadLanguage('eng+ell');
      await worker.initialize('eng+ell');

      // Configure OCR parameters for better Greek text recognition
      await worker.setParameters({
        tessedit_pageseg_mode: '6', // Assume uniform text block
        preserve_interword_spaces: '1',
        tessedit_char_whitelist: 'ΑΆΒΓΔΕΈΖΗΉΘΙΊΪΚΛΜΝΞΟΌΠΡΣΤΥΎΫΦΧΨΩΏαάβγδεέζηήθιίϊκλμνξοόπρσςτυύϋφχψωώ' +
          'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.,!@#$%^&*()_+-=[];:\'"\\|/<>?~ ',
        tessedit_enable_doc_dict: '0',  // Disable document dictionary
        tessedit_ocr_engine_mode: '3',  // Use Legacy + LSTM mode for better accuracy
        language_model_penalty_non_dict_word: '0.5',  // Lower penalty for non-dictionary words
        language_model_penalty_non_freq_dict_word: '0.5',
        textord_heavy_nr: '1',  // Handle text with variable spacing
        lstm_choice_mode: '2',  // Use more aggressive LSTM choices
        lstm_choice_iterations: '10'  // Increase LSTM iterations for better accuracy
      });

      OCRService.worker = worker;
      return OCRService.worker;
    } catch (error) {
      console.error('OCR Worker initialization failed:', error);
      throw error;
    } finally {
      OCRService.isInitializing = false;
    }
  }

  static async recognize(imageSrc) {
    if (!OCRService.worker) {
      throw new Error('OCR worker is not initialized');
    }

    try {
      // Pre-process recognition options
      const recognizeOptions = {
        rectangle: null, // Optional: Add specific region recognition if needed
        rotateAuto: true, // Enable automatic rotation detection
      };

      const { data } = await OCRService.worker.recognize(imageSrc, recognizeOptions);
      
      // Post-process the recognized text
      let text = data.text;
      
      // Clean up common OCR mistakes in Greek text
      text = text.replace(/([ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρστυφχψω])\s+([,.!;])/g, '$1$2')
                 .replace(/\s+/g, ' ')
                 .trim();

      return text;
    } catch (error) {
      console.error('OCR recognition failed:', error);
      throw error;
    }
  }

  static async terminate() {
    if (OCRService.worker) {
      await OCRService.worker.terminate();
      OCRService.worker = null;
    }
  }
}

export default OCRService;

**Comment by Balearica** [comment]
Can you provide a minimal, reproducible example of the issue?  While this OP does contain code, it is not something that can be easily run by other users, and no example images are provided.  You are most likely to get a response if you can replicate the issue using the [example code](https://github.com/naptha/tesseract.js/blob/master/examples/browser/basic-efficient.html) (modified as needed to replicate the issue) and providing image(s).

**Reply by RoseSamaras** [reply]
Hello Balearica 

Thank you for your quick reply, now it actually works, but the results are very poor, you can see in the image bellow

![Screenshot 2024-10-25 at 12 46 45](https://github.com/user-attachments/assets/060db65b-95c0-4dce-84a2-a9394a2b9609)

Can we get on a call, to show you the problem and see how we can collaborate?

Thank you!!

**Comment by YASSINE34501** [comment]
A few things in the posted code will each hurt on their own, and they compound. Roughly in order of how much they are likely costing you:

**1. The worker setup looks like the v2/v3 API, which would explain Greek not being read at all.**

```js
const worker = await createWorker({ langPath: '/tessdata', logger: m => console.log(m) });
await worker.loadLanguage('eng+ell');
await worker.initialize('eng+ell');
```

From v4 onward the signature is `createWorker(langs, oem, options)`, and `loadLanguage()` / `initialize()` were removed. So if you are on v5 or later, that object is being taken as the **first positional argument** — the language — rather than as options. `langPath` is never applied, and `eng+ell` is never actually loaded, because the two calls that would have loaded it no longer exist. The worker falls back to its default language and default asset path, which looks exactly like "it does not read Greek".

Check your installed version first, because if this is what is happening the rest barely matters. The v4+ form is:

```js
const worker = await createWorker(['eng', 'ell'], undefined, {
  langPath: '/tessdata',
  logger: m => console.log(m),
});
```

**2. `tessedit_ocr_engine_mode: '3'` asks for Legacy + LSTM.** The standard traineddata used here has no legacy component, so requesting the legacy engine gives you degraded or failed recognition rather than the "better accuracy" the comment suggests. Leave OEM at the default.

**3. `tessedit_char_whitelist` is a legacy-engine feature.** With the LSTM engine it generally makes output worse rather than better, and a whitelist this broad is not constraining anything useful anyway. I would remove it entirely and look at the unconstrained output first.

**4. Since you are self-hosting `langPath`, confirm both files actually resolve.** Open devtools → Network while the worker starts and check that `/tessdata/eng.traineddata.gz` **and** `/tessdata/ell.traineddata.gz` both return 200. This is worth the two minutes: I run tesseract.js self-hosted in production across five languages, and we shipped a bug where two languages were selectable in the UI whose model files had never been added to the download step. They 404'd against our own origin and recognition simply failed for those two — with a self-hosted `langPath` there is no CDN to quietly fall back to. The symptom presents as "this language does not work", not as "a file is missing", which is why it is worth checking explicitly rather than assuming.

**5. `PSM 6` (uniform text block) is probably the wrong mode for business cards.** Cards are usually multi-column with sparse, non-uniform text. Try PSM 3 (auto) or PSM 11 (sparse text) and compare on the same image.

**6. The preprocessing may be removing information rather than adding it.** You take a JPEG screenshot, upscale it 2x on canvas, apply contrast/brightness, then re-encode to JPEG at 0.95. Upscaling adds no detail, and each JPEG round trip adds artefacts around glyph edges — which is precisely where the recognizer is looking. Try passing the least-processed image you can get, and add steps back only if they measurably help.

One note on expectations: mixed Greek and Latin on the same card in a single pass is genuinely one of the harder cases, and small decorative type on business cards is hard even in one script. Getting (1) and (2) right should move you from "does not read Greek" to "reads Greek imperfectly", which is a different and much more tractable problem.
